expose-utility-functions-to-states.md

```diff
@@ -0,0 +1,133 @@
+exposing dubiously-useful functions to the CLI and template context.
+
+# Design
+[design]: #detailed-design
+
+## Option 1: Augmentation of LazyLoader
+
+An optional module-level dunder could be added to remote-execution modules, to
+suppress availability of utility functions. For example:
+
+```python
+__utility_funcs__ = ('foo', 'bar')
+```
+
+Any function name found in `__utility_funcs__` would be ignored by the loader
+by default, but they could be conditionally loaded by passing an argument when
+instantiating the `LazyLoader` class. This would allow `salt.loader.states()`
+to instantiate its own copy of the `minion_mods` loader that contains the
+utility functions, while the CLI and template context get a copy with them
+absent, preventing them from being exposed to end users.
+
+## Option 2: Move the utility functions to a new loader type
+
+This solution is far less elegant and more intrusive, but I'm including it as
+an alternative. It was my first idea for solving this problem, but I believe
+that [Option 1](#option-1-augmentation-of-lazyloader) is the superior solution.
+
+Utility functions would be moved to a different module, which would be in a
+directory processed by a loader instance. That loader would be added to the
+dunders which get packed onto the state modules.  Meanwhile, the execution
+modules can import these utility functions directly.
+
+The obvious first choice would be to use `__utils__` for this purpose, but I
+noticed there is already [a SEP to remove the `__utils__`
+loader](https://github.com/saltstack/salt-enhancement-proposals/pull/66). So,
+a different directory then? Assuming this directory is called `foo`, then you
+would for example have a `salt/modules/aptpkg.py` and a `salt/foo/aptpkg.py`.
+
+Splitting the code into separate modules for remote-execution and utility
+functions would appear at first to result in a duplicated `__virtual__`
+function, but objects beginning and ending in double-underscores are not
+name-mangled and thus can be imported. So, the `__virtual__` can be defined in
+the utility module and imported into the remote-execution module.
+
+## Alternatives
+[alternatives]: #alternatives
+
+This SEP already contains two potential implementations. The alternative to
+implementing this SEP would be to duplicate the logic from the `__virtual__` at
+every location where utility functions need to be used, which is an open
+invitation for the two (or more) copies of that logic to drift from one
+another.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+- For [Option 1](#option-1-augmentation-of-lazyloader), Salt once had a test
+  which parsed all docstrings in every execution module, ensuring that CLI
+  examples were included. I'm not sure if this test still exists, but utility
+  functions would not need CLI examples and this test case (if still present)
+  would need to be updated to account for this fact.
```

Comment:
```suggestion
- For [Option 1](#option-1-augmentation-of-lazyloader), Salt once had a test
  which parsed all docstrings in every execution module, ensuring that CLI
  examples were included. I'm not sure if this test still exists, but utility
  functions would not need CLI examples and this test case (if still present)
  would need to be updated to account for this fact.
```
There isn't a test case, but there is a pre-commit hook.

Reply:
No longer a concern with my proposed implementation from https://github.com/saltstack/salt-enhancement-proposals/pull/70#issuecomment-1625554331, since the utility functions now remain prefixed with underscores.